server/src/main/java/org/elasticsearch/common/util/CancellableSingleObjectCache.java

```diff
@@ -242,6 +252,31 @@ protected void closeInternal() {
             // Complete the future (and hence all its listeners) with an exception if it hasn't already been completed.
             future.onFailure(new TaskCancelledException("task cancelled"));
         }
+
+        public boolean supersedeIfStale() {
+            final CachedItem currentCachedItem = currentCachedItemRef.get();
+            if (currentCachedItem == this) {
+                // this item is still the freshest
+                return false;
+            }
+
+            if (currentCachedItem == null) {
+                // this item was superseded but the newer item was cancelled so we must proceed with a refresh for this item
+                return false;
+            }
+
+            cancellationChecks.runAll();
+            if (tryIncRef()) {
+                try {
+                    return currentCachedItem.addListener(future, () -> hasReferences() == false);
+                } finally {
+                    decRef();
+                }
+            } else {
+                // this item was cancelled, not superseded - return value doesn't really matter, refresh() will check for cancellation.
```

Comment:
I believe the future will already be completed with a `TaskCancelledException` inside `closeInternal` but yes that's something of an implementation detail. Adjusted in b5c6bf161a9faf3c3c9738e8d53a91058ddd380a.

Reply:
Ah, I missed that, thanks. In that case, I suppose it was more appropriate to return true here, since no further action is needed and any work done by the refresher will be wasted?